ckan/logic/action/create.py

```diff
@@ -166,13 +168,70 @@ def package_create(context, data_dict):
+            config.get('ckan.rename_dataset_name_when_collision', 'false'))
+
+    if 'name' in data_dict:
+        original_data_name = data_dict['name']
+        if rename_dataset_name_when_collision:
+            # modify name for shorten its length in 2 characters
+            # because we will check for a suffix up to 99
+            data_dict['name'] = data_dict['name'][:(PACKAGE_NAME_MAX_LENGTH-2)]
+
     data, errors = lib_plugins.plugin_validate(
         package_plugin, context, data_dict, schema, 'package_create')
     log.debug('package_create validate_errs=%r user=%s package=%s data=%r',
               errors, context.get('user'),
               data.get('name'), data_dict)
 
+    if rename_dataset_name_when_collision:
+        # checking for recoverable error (URL already in use)
+        # by adding a suffix
+        current_name_temp = ''
+        MAX_ATTEMPS = 99
+        url_in_use_msg = _(u'That URL is already in use.')
+        too_may_msg = _(u'Too many datasets with same name.')
+
+        if errors:
+            if ('name' in errors):
+                if (errors['name'][0] == url_in_use_msg):
+                    previous_data_name = data_dict['name']
+                    # search a new data name => previous name + counter
+                    i = 1
+                    # try MAX_ATTEMPS
+                    while i <= MAX_ATTEMPS:
+                        # get a new unused data name
+                        current_name_temp = data.get('name') + str(i)
+                        data_dict['name'] = current_name_temp
+                        data2, errors2 = lib_plugins.plugin_validate(
+                           package_plugin, context, data_dict, schema,
+                           'package_create')
+                        if errors2:
+                            if ('name' in errors2):
+                                if (errors2['name'][0] == url_in_use_msg):
+                                    if i >= (MAX_ATTEMPS-1):
+                                        # too many attemps => resign
+                                        model.Session.rollback()
+                                        raise df.Invalid(too_may_msg)
+                                    else:
+                                        i = i + 1
+                                else:
+                                    # there are other errors despite of
+                                    # changing its name => resign
+                                    i = MAX_ATTEMPS+1
+                            else:
+                                # there are other errors despite of
+                                # changing its name => resign
+                                i = MAX_ATTEMPS+1
+                        else:
+                            data['name'] = current_name_temp
+                            data_dict['name'] = current_name_temp
+                            errors = errors2
+                            i = MAX_ATTEMPS+1
+
```

Comment:
looping on the validation this way will be very slow on some sites.

How about updating the name validator to include this logic instead? It could make a single LIKE query to the DB to collect names with the same prefix then return the next available one. That should be a lot less code and perform much better.

Reply:
@wardi You're right. Doing this stuff in validator (using DB) is a much better way.
I've updated my code with this approach.
Thanks for your help!